delta/plugins/project-deletion/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/projectdeletion/ProjectDeletionPlugin.scala

```diff
@@ -0,0 +1,164 @@
+      config.includedProjects.exists(regex => regex.matches(pr.value.ref.toString))
+    }
+
+    def notExcluded(pr: ProjectResource): Boolean = {
+      !config.excludedProjects.exists(regex => regex.matches(pr.value.ref.toString))
+    }
+
+    def deletableDueToDeprecation(pr: ProjectResource): Boolean = {
+      config.deleteDeprecatedProjects && pr.deprecated
+    }
+
+    def deletableDueToBeingIdle(pr: ProjectResource, now: Instant): Task[Boolean] = {
+      projectIsIdle(pr, now) <&&> resourcesAreIdle(pr, now)
+    }
+
+    def projectIsIdle(pr: ProjectResource, now: Instant) = {
+      (now diff pr.updatedAt).toSeconds > config.idleInterval.toSeconds
+    }
+
+    def resourcesAreIdle(pr: ProjectResource, now: Instant) = {
+      lastEventTime(pr).map(_.isBefore(now.minus(Duration.ofMillis(config.idleInterval.toMillis))))
+    }
+
+    def allProjectsNotAlreadyDeleted: IO[Nothing, Seq[ProjectResource]] = {
+      allProjects
+        .map(_.filter(!_.value.markedForDeletion))
+    }
+
+    def shouldBeDeleted(now: Instant): ProjectResource => Task[Boolean] = { pr =>
+      {
+        (isIncluded(pr) && notExcluded(pr)) <&&> (deletableDueToDeprecation(pr) <||> deletableDueToBeingIdle(pr, now))
+      }
+    }
+
+    def deleteProjects(projects: Seq[ProjectResource]): Task[Unit] = {
+      Task
+        .traverse(projects)(
+          deleteProject(_).void.mapError(e => new RuntimeException(s"error deleting project from plugin: ${e}"))
+        )
+        .void
+    }
+
+    for {
+      allProjects      <- allProjectsNotAlreadyDeleted
+      now              <- IOUtils.instant
+      projectsToDelete <- allProjects.filterA[Task](shouldBeDeleted(now))
+      _                <- deleteProjects(projectsToDelete)
+    } yield {
+      ()
+    }
+  }
+
+  /**
+    * Constructs a stoppable ProjectDeletion process that is started in the background.
+    */
+  def started(
+      projects: Projects,
+      config: ProjectDeletionConfig,
+      projectStatistics: ProjectsStatistics
+  ): Task[ProjectDeletionPlugin] = {
+
+    def lastEventTime(pr: ProjectResource): Task[Instant] = {
+      projectStatistics.get(pr.value.ref).flatMap {
+        case Some(stats) => IO.pure(stats.lastEventTime)
+        case None        => IO.raiseError(new IllegalArgumentException(s"Project '${pr.value.ref}' was not found."))
+      }
+    }
+
+    val allProjects: UIO[Seq[ProjectResource]] = {
+      projects
+        .list(
+          Pagination.OnePage,
+          ProjectSearchParams(filter = _ => UIO.pure(true)),
+          Ordering.by(_.updatedAt) // this is not needed, we are forced to specify an ordering
+        )
+        .map(_.results)
+        .map(_.map(_.source))
+    }
+
+    def deleteProject(pr: ProjectResource): IO[ProjectRejection, Unit] = {
+      implicit val caller: Subject = Identity.Anonymous
+      projects
+        .delete(pr.value.ref, pr.rev)
+        .void
+    }
+
+    val continuousStream = Stream
```

Comment:
I forgot about it but we need the supervisor here to run the stream so that it gets restarted if it fails (due for example to a Network error)

Reply:
Yeah this is currently done in a weird way by using recover and passing back the original stream.. Do you have an example of where we use a supervisor I can look at?

Reply:
One simple example is `EventMetricsProjection`. Or you can look for usages of the `Supervisor.run` function